### ISSUE TITLE
Unable to follow an Owncast account from Misskey

### ISSUE THREAD

**Issue by gabek** [owncast/owncast#1690]
I filed an issue with Misskey in hopes they can help me troubleshoot what's going on.

https://github.com/misskey-dev/misskey/issues/8165

**Comment by gabek** [comment]
Update: It's the same issue PeerTube originally had and had to resolve regarding them not supporting the modern best practices of `hs2019` http signatures.

The good news is there's nothing on Owncast's side that needs to be changed. In theory it'll just start working if they update on their side.

Closing this, since there's nothing to be done here.

**Comment by ghost** [comment]
I'm not very familiar with ActivityPub so I might be missing something here, but I'm looking at `Accept` activities from pleroma and owncast, and I think I figured out what is breaking this (after updating for the `hs2019` signatures)

pleroma's Accept activity:
```json
{
    "@context": [
      "https://www.w3.org/ns/activitystreams",
      "https://pleroma.nya/schemas/litepub-0.1.jsonld",
      [Object]
    ],
    "actor": "https://pleroma.nya/users/catgirl",
    "cc": [],
    "id": "https://pleroma.nya/activities/52105405-3700-42ae-a9eb-1f1ca847abe0",
    "object": {
      "actor": "https://misskey.nya/users/8yq0onxxid",
      "id": "https://misskey.nya/14692b04-b86b-4dc1-8189-8e51af701720",
      "object": "https://pleroma.nya/users/catgirl",
      "type": "Follow"
    },
    "to": [ "https://misskey.nya/users/8yq0onxxid" ],
    "type": "Accept"
}
```

owncast's Accept activity:
```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "actor": "https://owncast.nya/federation/user/catgirl",
  "id": "https://owncast.nya/federation/V4JX4mU7g",
  "object": "https://misskey.nya/d9c64a9c-e359-4f03-94d3-8470cbc50507",
  "type": "Accept"
}
```

the important difference between these two is that pleroma sends the `object` key as an object, so misskey doesn't try to resolve it; owncast sends the `object` key as a string, so misskey tries to resolve it, which won't work since that's not actually a valid URL that points to anything.
I'm not sure if that URL is correct or what it's supposed to be, so I don't really know how this should be fixed. just thought I'd point out the little bit of info I was able to figure out ^^

**Comment by gabek** [comment]
Thanks to @spiralw's help we were able to determine there were two pieces at play:

1. Misskey doesn't like sending unresolved IRIs as the object even though Mastodon and Pleroma are fine with it.
2. Misskey still hasn't merged in the `hs2019` update upstream.

However testing on a Misskey server that _does_ have the `hs2019` patch _and_ sending along the full object payload instead of an IRI allows for a successful follow as seen here:

![image](https://user-images.githubusercontent.com/414923/162521854-f1c30005-b852-447b-8085-7433af2343d9.png)

**Comment by gabek** [comment]
I'll just keep pinging that Misskey issue in hopes they release the fix.

**Comment by Johann150** [comment]
Hi just checking in from the Misskey side, do I understand correctly that this fix is not yet released? Because I think I just got this error while testing https://github.com/misskey-dev/misskey/pull/8635#issuecomment-1125355809

And I think I'll look into this bug too on our side because if I saw correctly the `object` seems to be like `https://example.com/<some uuid>` which is to my knowledge a fallback behaviour. Maybe we forgot to set a proper ID on our side in the original follow request 😅

**Comment by gabek** [comment]
As far as I know there's two different issues at play.

1. Misskey doesn't accept IRIs as actors for some reason. So instead of fighting Misskey on this issue I changed how Owncast sends actors to use the full object instead. This change to Owncast will be released in 0.0.12. If you'd like to test against a server that's running 0.0.12 you can follow `goodnight@nightly.owncast.tv` (https://nightly.owncast.tv).
2. Misskey didn't support `hs2019`.

@spiralw was able to successfully follow an Owncast instance from Misskey using the Owncast `develop` branch and the using the `hs2019` update. Because he was able to get it to work (I don't personally have a dev Misskey environment) it seems like things are resolved but pending.

<blockquote><img src="https://nightly.owncast.tv/logo/external" width="48" align="right"><div><img src="/img/favicon/favicon-16x16.png" height="14"> Owncast Nightly</div><div><strong><a href="https://nightly.owncast.tv/">Owncast Nightly</a></strong></div><div>This is a server running the nightly build of Owncast.</div></blockquote>

**Comment by Johann150** [comment]
Yes as I said I tried to test my current fix for the Misskey side so # 2 should not be a problem in that case. But my question is still answered that the fix on your side is not released or at least not in 0.0.11 which I was testing against.

Testing against the one you provided works with my current open PR linked above. 👍

**Comment by gabek** [comment]
Hurray!  That's great to hear.

One of the reasons 0.0.12 wasn't released yet was I wanted to be able to test against real Misskey servers in the wild and make sure things were working, but I wasn't able to do that. But I certainly feel better now. Thanks!
